ucp/benchmarks/backends/ucp_core.py

```diff
@@ -0,0 +1,287 @@
+    worker: ucp._libs.UCXWorker
+        UCX-Py core Worker object where to register the allocator.
+    """
+    if not args.enable_am:
+        return
+
+    import numpy as np
+
+    worker.register_am_allocator(
+        lambda n: np.empty(n, dtype=np.uint8), ucx_api.AllocatorType.HOST
+    )
+
+    if args.object_type == "cupy":
+        import cupy as cp
+
+        worker.register_am_allocator(
+            lambda n: cp.empty(n, dtype=cp.uint8), ucx_api.AllocatorType.CUDA
+        )
+    elif args.object_type == "rmm":
+        import rmm
+
+        worker.register_am_allocator(
+            lambda n: rmm.DeviceBuffer(size=n), ucx_api.AllocatorType.CUDA
+        )
+
+
+class UCXPyCoreServer(BaseServer):
+    def __init__(self, args, xp, queue):
+        self.args = args
+        self.xp = xp
+        self.queue = queue
+
+    def run(self):
+        ctx = ucx_api.UCXContext(
+            feature_flags=(
+                ucx_api.Feature.AM
+                if self.args.enable_am is True
+                else ucx_api.Feature.TAG,
+            )
+        )
+        worker = ucx_api.UCXWorker(ctx)
+
+        register_am_allocators(self.args, worker)
+
+        # A reference to listener's endpoint is stored to prevent it from going
+        # out of scope too early.
+        ep = None
+
+        op_lock = Lock()
+        finished = [0]
+        outstanding = [0]
+
+        def op_started():
+            with op_lock:
+                outstanding[0] += 1
+
+        def op_completed():
+            with op_lock:
+                outstanding[0] -= 1
+                finished[0] += 1
+
+        def _send_handle(request, exception, msg):
+            # Notice, we pass `msg` to the handler in order to make sure
+            # it doesn't go out of scope prematurely.
+            assert exception is None
+            op_completed()
+
+        def _tag_recv_handle(request, exception, ep, msg):
+            assert exception is None
+            req = ucx_api.tag_send_nb(
+                ep, msg, msg.nbytes, tag=0, cb_func=_send_handle, cb_args=(msg,)
+            )
+            if req is None:
+                op_completed()
+
+        def _am_recv_handle(recv_obj, exception, ep):
+            assert exception is None
+            msg = Array(recv_obj)
+            ucx_api.am_send_nbx(
+                ep, msg, msg.nbytes, cb_func=_send_handle, cb_args=(msg,)
+            )
+
+        def _listener_handler(conn_request, msg):
+            global ep
+            ep = ucx_api.UCXEndpoint.create_from_conn_request(
+                worker,
+                conn_request,
+                endpoint_error_handling=True,
+            )
+
+            # Wireup before starting to transfer data
+            if self.args.enable_am is True:
+                ucx_api.am_recv_nb(ep, cb_func=_am_recv_handle, cb_args=(ep,))
+            else:
+                wireup = Array(bytearray(len(WireupMessage)))
+                op_started()
+                ucx_api.tag_recv_nb(
```

Comment:
I meant that should we not wait on the request to complete (possibly before starting the rest of the ping-pong messages).

The client does a blocking send/recv to exchange the wireup information, but on the server we do a nonblocking wireup.

Reply:
I see what you're saying now. I think this is not a problem to have the synchronization only on the client, given this is also a Tag message as we have a guarantee of message ordering. The case where this may be problematic is if the message fails, but then it would anyway eventually be reported in order as well.

Do you have any other concerns that I'm perhaps not thinking about?

Reply:
Nope.